digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/CommonHistoryFiller.scala

```diff
@@ -109,6 +121,56 @@ object CommonHistoryFiller {
     }.values.flatten.toSeq
   }
 
+  private def setCalibrationPoints(roadAddresses: Seq[RoadAddress]): Seq[RoadAddress] = {
+    def makeStartCP(roadAddress: RoadAddress) = {
+      Some(CalibrationPoint(roadAddress.linkId, if (roadAddress.sideCode == TowardsDigitizing)
+        0.0 else
+        GeometryUtils.geometryLength(roadAddress.geometry),
+        roadAddress.startAddrMValue))
+    }
+
+    def makeEndCP(roadAddress: RoadAddress) = {
+      Some(CalibrationPoint(roadAddress.linkId, if (roadAddress.sideCode == AgainstDigitizing)
+        0.0 else
+        GeometryUtils.geometryLength(roadAddress.geometry),
+        roadAddress.endAddrMValue))
+    }
+
+    def fillCPs(roadAddress: RoadAddress, both: Boolean = false, putCPAtStart: Boolean = false): RoadAddress = {
+      val startCP = makeStartCP(roadAddress)
+      val endCP = makeEndCP(roadAddress)
+      if (both) {
+        roadAddress.copy(calibrationPoints = (startCP, endCP))
+      } else {
+        if (!putCPAtStart) {
+          roadAddress.copy(calibrationPoints = (None, endCP))
+        } else {
+          roadAddress.copy(calibrationPoints = (startCP, None))
+        }
+      }
+    }
+
+    val startNeedsCP = roadAddresses.head

```

Comment:
Did you tests the case where you have a empty roadAddresses. I think you will get the following error "head of empty list".